### PR TITLE
fx quant: support for ndim followed by binary op

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53875 fx quant: remove unnecessary dequants for getattr
* #53614 fx quant: clean up names of quantize handlers
* #53196 fx quant: ensure fp32 sum works correctly with quantized layers
* #53187 fx quant: fix using size of quant layer in torch._assert
* **#53120 fx quant: support for ndim followed by binary op**

Summary:

Currently there is a pattern which is not handled correctly by
FX graph mode quantization:

```
def forward(self, x):
    ndim = x.ndim
    # or add, mul, div, etc
    x = torch.sub(x, ndim)
    return x
```

The reason this does not work is as follows:
1. x.ndim becomes a getattr node
2. the real world type of x.ndim is an integer, but this is not known from the graph (yet)
3. binary ops such as `torch.sub` require quantization of inputs
4. the framework inserts an observer to observe the output of `ndim`
5. the observer fails because `ndim` is not a Tensor

The fix implemented in this PR is adding a heuristic on whether a node has all of its args
known to be not tensors.  In the longer term, this may need to be improved to adding reasoning
about the return type of various ops, propagating it through the graph, and updating the rules to use it.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_getattr_with_nontensor_result
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26756180](https://our.internmc.facebook.com/intern/diff/D26756180)

### CODE REVIEW

File: torch/quantization/fx/quantization_patterns.py [hunk 1]
```diff
@@ -192,9 +192,23 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
...
                 kwargs = {**self.binary_op_node.kwargs}
-                add_args = (*load_arg(quantized=True)(self.binary_op_node.args), scale_arg, zero_point_arg)
-                op = quantizer.quantized_graph.create_node(
-                    'call_function', self.quantized_binary_op, add_args, kwargs)
+                # not for land, just a hack to unblock for now
+                # TODO(before land): real fix
+                # add_args = (*load_arg(quantized=True)(self.binary_op_node.args), scale_arg, zero_point_arg)
+                # op = quantizer.quantized_graph.create_node(
+                    # 'call_function', self.quantized_binary_op, add_args, kwargs)
+                if True:
```
Comment (vkuzo) [inline torch/quantization/fx/quantization_patterns.py:200]: We should figure out the best implementation here.  A challenge with the current approach is that there is a new type of arg, "potentially quantized".  So, we either need to increase the # of types of args from two to three, or think of a different approach to this PR.

File: torch/quantization/fx/quantize.py [hunk 2]
```diff
@@ -318,6 +318,60 @@ def node_arg_is_bias(node: Node, arg: Any) -> bool:
                     return True
     return False
 
+def all_node_args_have_no_tensors(node: Node) -> bool:
```
Comment (vkuzo) [inline torch/quantization/fx/quantize.py:321]: in the short term, this sounds OK.  In the longer term, we might need to evaluate introducing an output type for nodes, which is known for some nodes (and maybe not others) and can propagate through various ops.  For example, we know that the output of `x.size(...)` is always an integer, etc.

**Review by jerryzh168 — APPROVED** [review]
The try except part looks hacky, but the rest seems to be OK I think.

File: torch/quantization/fx/quantization_patterns.py (hunk 1 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantization_patterns.py:200]: yeah this part looks hacky, is it possible to identify that a node is produced by ndim, size etc. and filter out these nodes during loading? something like:

```
int_node_idxs = [1]
all_arg_idxs = [1, 2]
quantized_idxs = [2]
... = load_arg(quantized=quantized_idxs)...
```

**Review by jerryzh168 — APPROVED** [review]

**Review by raghuramank100 — APPROVED** [review]

File: torch/quantization/fx/utils.py [hunk 3]
```diff
@@ -330,3 +330,46 @@ def create_qparam_nodes(quantizer: QuantizerCls, node_name: str, scale: Any, zer
...
+    for arg in node.args:
+        if isinstance(arg, list):
+            for list_el in arg:
+                if isinstance(list_el, Node):
+                    this_list_el_args_have_no_tensors = \
+                        all_node_args_have_no_tensors(list_el)
+                    found_one_tensor = found_one_tensor or \
+                        (not this_list_el_args_have_no_tensors)
+        elif isinstance(arg, int):
+            pass
```
Comment (jerryzh168) [inline torch/quantization/fx/utils.py:367]: pass? should we remove?

File: torch/quantization/fx/utils.py [hunk 4]
```diff
@@ -330,3 +330,46 @@ def create_qparam_nodes(quantizer: QuantizerCls, node_name: str, scale: Any, zer
...
+        else:
+            if isinstance(arg, Node):
+                this_arg_args_have_no_tensors = all_node_args_have_no_tensors(arg)
+                found_one_tensor = found_one_tensor or \
+                    (not this_arg_args_have_no_tensors)
+            else:
```
Comment (jerryzh168) [inline torch/quantization/fx/utils.py:373]: why is this True?

Comment (vkuzo) [reply torch/quantization/fx/utils.py:373]: yeah, this is a heuristic, we are assuming all "unknown" types are tensors

btw, in the future, we can delete this code and switch to https://github.com/pytorch/pytorch/pull/53653, once it is ready

**Pull request closed without merging**